compiler/hash-parser/src/parser/macros.rs

```diff
@@ -0,0 +1,153 @@
+//! Parsing code for various kind of macros.
+
+use hash_ast::ast::{
+    AstNode, MacroInvocation, MacroInvocationArg, MacroInvocationArgs, MacroInvocations, MacroKind,
+    Name,
+};
+use hash_token::{delimiter::Delimiter, Token, TokenKind};
+use hash_utils::thin_vec::thin_vec;
+
+use super::AstGen;
+use crate::diagnostics::error::{ParseErrorKind, ParseResult};
+
+impl<'stream, 'resolver> AstGen<'stream, 'resolver> {
+    fn parse_macro_prefix(&self, kind: MacroKind) -> Option<()> {
+        match kind {
+            MacroKind::Token if self.parse_token_fast(TokenKind::At).is_some() => {
+                self.skip_token();
+                Some(())
+            }
+            MacroKind::Ast if self.parse_token_fast(TokenKind::Pound).is_some() => {
+                self.skip_token();
+                Some(())
+            }
+            _ => None,
+        }
+    }
+
+    fn parse_macro_arg(&mut self) -> ParseResult<AstNode<MacroInvocationArg>> {
```

Comment:
what's the point of this being separate from function call args

Reply:
If the two are to be merged, then we should probably rename `ConstructorCallArg` to something more general. Any ideas?

Reply:
Maybe just Arg? It can even be unified with TyArg

Reply:
Actually, `ConstructorCallArg` is defined as https://github.com/hash-org/hashc/blob/7a4f43d1c3d695d62e2e831699edd4b48ebb0dbd/compiler/hash-ast/src/ast.rs#L1864-L1873 which contains a macro invocation in the `ConstructorCallArg` itself whilst `MacroInvocationArg` does not. I guess the distinction is that a macro arg is not allowed to have inner args? But, we could still use the same variant... I don't know which is better.

Reply:
Also, does this mean that `TyArg` needs `macro_args` as well?

Reply:
Maybe there can be a EnrichedArg which is an Arg with an optional macro. And the former is what is used everywhere other than macros

Reply:
And yes, which is another reason to make this change

Reply:
However, `TyArg` has a value as `Ty` not `Expr` so it isn't possible to unify them like that.

Reply:
I do think that `ConstructorCallArg` should be renamed to `Arg` tho.

Reply:
Indeed, though it could trivially be wrapped in Expr::Ty. It is a cost of an extra ast node though, so perhaps settle for ExprArg/TyArg

Reply:
This means that we still need a `MacroArg` variant.

Reply:
Looks good, though I think `s/macro_args/macros/`